x/staking/src/keeper/tx.rs

```diff
@@ -308,4 +401,89 @@ impl<
         //     }, nil
         Ok(())
     }
+
+    /// undelegate_cmd_handler defines a method for performing an undelegation from a delegate and a validator
+    pub fn undelegate_cmd_handler<DB: Database>(
+        &self,
+        ctx: &mut TxContext<'_, DB, SK>,
+        msg: &UndelegateMsg,
+    ) -> Result<(), AppError> {
+        let shares = self
+            .validate_unbond_amount(
+                ctx,
+                &msg.delegator_address,
+                &msg.validator_address,
+                msg.amount.amount,
+            )
+            .map_err(|e| AppError::Custom(e.to_string()))?;
+
+        let params = self.staking_params_keeper.try_get(ctx)?;
+        if msg.amount.denom != params.bond_denom {
+            return Err(AppError::InvalidRequest(format!(
+                "invalid coin denomination: got {}, expected {}",
+                msg.amount.denom, params.bond_denom
+            )));
+        }
+
+        let completion_time = self
+            .undelegate(ctx, &msg.delegator_address, &msg.validator_address, shares)
+            .map_err(|e| AppError::Custom(e.to_string()))?;
+
+        // TODO
```

Comment:
I think that you may omit telemetry completely

Reply:
removed